Make Stream.Flush no-op on read-only streams

martincostello opened the pull request:
# Make Stream.Flush no-op

Make `Stream.Flush{Async}()` implementations a no-op.

## Description

Make implementations of `Stream.Flush{Async}()` a no-op instead of throw an exception.

Resolves #57064.

martincostello committed:
Make Stream.Flush no-op

Make implementations of `Stream.Flush()` a no-op instead of throw an exception.
Relates to #57064.

teo-tsirpanis commented:
You could override `FlushAsync()` to return a completed task, to avoid scheduling to run `Flush()` on the thread pool.

martincostello committed:
Make Stream.FlushAsync no-op

- Override `Stream.FlushAsync()` methods as a no-op.
- Make implementations of `Stream.FlushAsync()` a no-op instead of throw an exception.

martincostello commented:
Done - thanks, that flushed (no pun intended) out two `FlushAsync()` methods that were throwing that my initial search missed.

amcasey commented on src/Components/Server/src/Circuits/RemoteJSDataStream.cs:
```diff
@@ -161,7 +161,11 @@ public override long Position
     public override void Flush()
-        => throw new NotSupportedException();
+    {
+        // No-op
```
```suggestion
        // No-op - Flush is specifically allowed on read-only streams.
```

martincostello replied:
Do you want me to update all of these, or are you not fussed with spinning CI to update the comments?

amcasey replied:
I can live without the comment(s).

amcasey commented on src/Shared/SegmentWriteStream.cs:
```diff
@@ -99,12 +99,11 @@ protected override void Dispose(bool disposing)
     {
-        if (!CanWrite)
-        {
-            throw new ObjectDisposedException(nameof(SegmentWriteStream), "The stream has been closed for writing.");
```
This feels a little different - does this only happen after the stream is closed?

amcasey replied:
Personally, I like the exception since I think flushing (or doing anything else) after Dispose is a programming error, but I see that `CryptoStream` itself [doesn't](https://source.dot.net/#System.Security.Cryptography/System/Security/Cryptography/CryptoStream.cs,203) throw under those circumstances, so I'm okay with not throwing here either.

amcasey commented on src/Servers/HttpSys/src/RequestProcessing/RequestStream.cs:
```diff
@@ -77,10 +77,12 @@ public override long Seek(long offset, SeekOrigin origin)
 
     public override void SetLength(long value) => throw new NotSupportedException(Resources.Exception_NoSeek);
 
-    public override void Flush() => throw new InvalidOperationException(Resources.Exception_ReadOnlyStream);
```
Was this the only consumer of `Resources.Exception_ReadOnlyStream`?

martincostello replied:
Good catch, it does appear to be unused after this change.

amcasey replied:
Unresolving, in case you were using it to track work.

martincostello replied:
Thanks - turns out I was wrong, it's still used here:

https://github.com/dotnet/aspnetcore/blob/43daa49193408a047e1df3fe1fc44dea80c4b8ca/src/Servers/HttpSys/src/RequestProcessing/RequestStream.cs#L302-L315

amcasey reviewed (commented):
Please double-check SegmentWriteStream.  Otherwise, LGTM.
